graylog2-shared/src/main/java/org/graylog2/shared/plugins/PluginLoader.java

```diff
@@ -39,12 +48,17 @@ public PluginLoader(File pluginDir) {
     }
 
     public Set<Plugin> loadPlugins() {
-        final ImmutableSet.Builder<Plugin> plugins = ImmutableSet.builder();
+        final Set<Plugin> sorted = ImmutableSortedSet.orderedBy(new PluginComparator())
+                .addAll(loadClassPathPlugins())
+                .addAll(loadJarPlugins())
+                .build();
 
-        plugins.addAll(loadClassPathPlugins());
-        plugins.addAll(loadJarPlugins());
+        final ImmutableSet.Builder<Plugin> unique = ImmutableSet.builder();
+        for (Plugin plugin : sorted) {
+            unique.add(new PluginAdapter(plugin));
```

Comment:
The `PluginAdapter` class takes care of eliminating duplicate plugins (in different versions) by only taking the unique ID and the name of the plugins into account for `equals()` and `hashcode()`. The `sorted` set can still contain duplicates (as described in https://github.com/Graylog2/graylog2-server/pull/948#discussion_r24404738).